src/java.net.http/share/classes/jdk/internal/net/http/HttpClientImpl.java

```diff
@@ -138,24 +147,39 @@ public Thread newThread(Runnable r) {
     static final class DelegatingExecutor implements Executor {
         private final BooleanSupplier isInSelectorThread;
         private final Executor delegate;
-        DelegatingExecutor(BooleanSupplier isInSelectorThread, Executor delegate) {
+        private final BiConsumer<Runnable, Throwable> errorHandler;
+        DelegatingExecutor(BooleanSupplier isInSelectorThread,
+                           Executor delegate,
+                           BiConsumer<Runnable, Throwable> errorHandler) {
             this.isInSelectorThread = isInSelectorThread;
             this.delegate = delegate;
+            this.errorHandler = errorHandler;
         }
 
         Executor delegate() {
             return delegate;
         }
 
+
+
         @Override
         public void execute(Runnable command) {
             if (isInSelectorThread.getAsBoolean()) {
-                delegate.execute(command);
+                ensureExecutedAsync(command);
             } else {
                 command.run();
             }
         }
 
+        public void ensureExecutedAsync(Runnable command) {
+            try {
+                delegate.execute(command);
+            } catch (Throwable t) {
+                errorHandler.accept(command, t);
+                ASYNC_POOL.execute(command);
```

Comment:
Does this mean that we will now attempt to use an `ASYNC_POOL` even if a had supplied an `Executor`? The `CompletableFuture#defaultExecutor` which is what the `ASYNC_POOL` represents states:
```
Returns the default Executor used for async methods that do not specify an Executor.
```
which would now contradict with this code. With the error handler logic in place (one line above), do you think we should just give up on running the `command` instead of passing to the default executor?

Furthermore, if we do decide to use the default executor in cases like this, is that something that all implementations of the `HttpClient` are expected to do (i.e. would it be a specification) or is this more an implementation detail.

Reply:
1. Dependent actions added by the caller to the CF returned by the HttpClient are executed in the FJP
2. This is an exceptional case - we're doing this only when we're shutting down the HttpClient. I don't think we should document it - this is an implementation detail. But we should probably document that shutting down the executor while operations are still in progress can lead to unpredictable/unspecified behavior.

Note: If we had VirtualThreads we would probably create a new VirtualThread here, rather than using the FJP.

Reply:
Thank you for that clarification, especially this part which I wasn't aware of:

> 1. Dependent actions added by the caller to the CF returned by the HttpClient are executed in the FJP